composer/algorithms/distillation/distillation.py

```diff
@@ -0,0 +1,250 @@
+                return True
+            else:
+                return False
+        else:
+            return False
+
+    def apply(self, event: Event, state: State, logger: Logger) -> Optional[int]:
+
+        if event == Event.FIT_START:
+            # Handle parallelization
+            for teacher_i, teacher in enumerate(self.teachers):
+                if state.fsdp_config is not None:
+                    try:
+                        prepare_fsdp_module(teacher,
+                                            optimizers=None,
+                                            fsdp_config=state.fsdp_config,
+                                            precision=state.precision)
+                    except:
+                        log.error('Unable to prepare teacher model for FSDP')
+                        raise
+                # TODO: Figure out whether we even need DDP wrapping, given that teacher
+                # params don't receive gradients
+                elif state.fsdp_config is None and dist.get_world_size() > 1:
+                    try:
+                        self.teachers[teacher_i] = prepare_ddp_module(teacher, find_unused_parameters=True)
+                    except:
+                        log.error('Unable to prepare teacher model for DDP')
+                        raise
+                else:
+                    try:
+                        self._move_teacher_model_to_device(teacher.module, state.model, state.device)
+                    except:
+                        log.error('Unable to move teacher.module to device. Will attempt to move teacher.')
+                        try:
+                            self._move_teacher_model_to_device(teacher, state.model, state.device)
+                        except:
+                            log.error('Unable to move teacher to device.')
+                            raise
+
+        elif event == Event.AFTER_LOSS:
+
+            input = None
+            if isinstance(state.batch, dict):
+                state.batch.pop('labels')
+                input = state.batch
+            else:
+                input = state.batch
+
+            # Get teacher(s) to sample from
+            t_idx = np.random.choice(range(len(self.teachers)), size=self.n_teachers_to_sample, replace=False)
+            # get teacher output
+            teacher_outputs = []
+
+            # get teacher output
+            with torch.no_grad():
+                with get_precision_context(state.precision):
+                    for idx in t_idx:
+                        teacher_output = self.teachers[int(idx)](input)
+                        if isinstance(teacher_output, MaskedLMOutput):
+                            teacher_outputs.append(teacher_output.logits.detach().clone())
+                        else:
+                            teacher_outputs.append(teacher_output)
+            # Avg teacher outputs
+            teacher_output = torch.mean(torch.stack(teacher_outputs), dim=0)
```

Comment:
Is this the correct/desired formulation? 

E.g. is the KL_Div(student || avg_teacher) == avg(KL_DIV(student || teacher_i))? 

I don't think it is, but I could be wrong?

Reply:
But also trying to match a single distribution might be easier than matching 'N' distributions empirically -- at least intuitively. I don't know what's common best practice or if there is one though

Reply:
This is the desired formulation. I'm not sure if matching the individual distributions is equivalent, but I do know that calculating N teacher losses is going to be slower. In our experiments we have mostly been using MSE for multi teacher distillation anyways so it really doesn't matter.